src/hotspot/share/jfr/support/jfrThreadLocal.hpp

```diff
@@ -283,6 +284,55 @@ class JfrThreadLocal {
   void set_thread_blob(const JfrBlobHandle& handle);
   const JfrBlobHandle& thread_blob() const;
 
+  // CPU time sampling
+#ifdef LINUX
+private:
+
+ enum CPUTimeLockState {
+    UNLOCKED,
+    // locked for enqueuing
+    ENQUEUE,
+    // locked for dequeuing
+    DEQUEUE,
+    // locked for writing native event out of safepoint
+    OUT_OF_SAFEPOINT
+  };
+
+  bool _has_cpu_timer = false;
+  timer_t _cpu_timer;
+  volatile CPUTimeLockState _cpu_time_jfr_locked = UNLOCKED;
+  volatile bool _has_cpu_time_jfr_events = false;
+  volatile u4 _cpu_time_lost_samples = 0;
+  JfrLocalCPUTimeTraceStack _cpu_time_jfr_stack{100};
+  volatile bool _wants_cpu_time_out_of_safepoint_recording = false;
+
+public:
+  void set_timerid(timer_t timer);
+  void unset_timerid();
+  timer_t timerid() const;
+  bool has_timerid() const;
+
+  bool is_cpu_time_jfr_enqueue_locked();
+  bool is_cpu_time_jfr_dequeue_locked();
+  bool is_any_cpu_time_jfr_queue_lock_aquired();
+
+  bool acquire_cpu_time_jfr_enqueue_lock();
+  bool acquire_cpu_time_jfr_out_of_safepoint_lock();
+  void acquire_cpu_time_jfr_dequeue_lock();
+  void release_cpu_time_jfr_queue_lock();
```

Comment:
Why not just define cpu_time_jfr_lock()/cpu_time_jfr_unlock()? It’s kind of confusing reading the methods where these are used since we need to remember they are the same lock.

Reply:
They are using the same lock, but the lock has four different states:
- enqueue: for enqueuing a new trace
- out of safepoint: for storing into JFR from the sampler thread
- dequeue: for dequeuing and storing into JFR at a safepoint
- unlocked: the base state

It is only possible to move between the base state and another state, but not between base states. This ensures that the three operations don't overlap.